### PR TITLE
[WIP]Review for solution to check relative path of SDL.

### PR DESCRIPTION
**Pull request by istoimenova**
Please check solution for verifying SDL relative path.
In case you have other suggestions, please list them in comments and we clarify the best solution.

### CODE REVIEW

**Comment by istoimenova** [comment]
@ttdong, @OVikhrov, @LuxoftAKutsan, @VVeremjova, @maria-mih, @vvvakulenko, @aderiabin, @dboltovskyi 
All, please check described solution.

**Comment by istoimenova** [comment]
[WIP] - PR is not a script.

File: to_clarify/Solutions/solution_path_to_sdl.lua [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+-- (data.params.cmdIcon.value ~= value_Icon )
+--
+-- Example:
+-- compare string '/home/istoimenova/SmartDeviceLinkCore/test_run/SDL_bin/storage/0000001_12ca17b49af2289436f303e0166030a21e525d266e209267433801a8fd4071a0/icon.png' defined by (data.params.cmdIcon.value)
+-- with string '/home/istoimenova/SmartDeviceLinkCore/test_run/SDL_bin/storage/0000001_12ca17b49af2289436f303e0166030a21e525d266e209267433801a8fd4071a0/icon.png' defined by (value_Icon)
+--
+-- Described solution makes scripts running on CI and locally
+
+config.SDLStoragePath = commonPreconditions:GetPathToSDL() .. "storage/"
+local storagePath = config.SDLStoragePath..config.application1.registerAppInterfaceParams.appID.. "_" .. config.deviceMAC.. "/"
```
Comment (LuxoftAKutsan) [inline to_clarify/Solutions/solution_path_to_sdl.lua]: @istoimenova that is the place where you shoul use path.join.

File: to_clarify/Solutions/solution_path_to_sdl.lua [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+--
+-- (data.params.cmdIcon.value ~= value_Icon )
+--
+-- Example:
+-- compare string '/home/istoimenova/SmartDeviceLinkCore/test_run/SDL_bin/storage/0000001_12ca17b49af2289436f303e0166030a21e525d266e209267433801a8fd4071a0/icon.png' defined by (data.params.cmdIcon.value)
+-- with string '/home/istoimenova/SmartDeviceLinkCore/test_run/SDL_bin/storage/0000001_12ca17b49af2289436f303e0166030a21e525d266e209267433801a8fd4071a0/icon.png' defined by (value_Icon)
+--
+-- Described solution makes scripts running on CI and locally
+
+config.SDLStoragePath = commonPreconditions:GetPathToSDL() .. "storage/"
```
Comment (LuxoftAKutsan) [inline to_clarify/Solutions/solution_path_to_sdl.lua]: @istoimenova that is the place where you should use path.join.

Comment (LuxoftAKutsan) [inline to_clarify/Solutions/solution_path_to_sdl.lua]: @aderiabin please tell where is it.

File: to_clarify/Solutions/solution_path_to_sdl.lua [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+local storagePath = config.SDLStoragePath..config.application1.registerAppInterfaceParams.appID.. "_" .. config.deviceMAC.. "/"
+
+EXPECT_HMICALL("UI.AddCommand",
+{
+  cmdID = i,
+  menuParams = { parentID = 0, position = 0, menuName ="Commandpositive" .. tostring(i)}
+})
+:ValidIf(function(_,data)
+  local value_Icon = storagePath .. "icon.png"
```
Comment (LuxoftAKutsan) [inline to_clarify/Solutions/solution_path_to_sdl.lua]: @istoimenova path.join.

File: to_clarify/Solutions/solution_path_to_sdl.lua [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  if (string.match(data.params.cmdIcon.value, "%S*" .. "("..string.sub(storagePath, 2).."icon.png)" .. "%W*$") == nil )  and
```
Comment (LuxoftAKutsan) [inline to_clarify/Solutions/solution_path_to_sdl.lua]: @istoimenova such complex check I woul suggest to extract to local function with expressive name

File: to_clarify/Solutions/solution_path_to_sdl.lua [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+     (data.params.cmdIcon.value ~= value_Icon ) then
+    print("\27[31m value of cmduIcon is WRONG. Expected: ".. value_Icon .. "; Real: " .. data.params.cmdIcon.value .. "\27[0m")
```
Comment (LuxoftAKutsan) [inline to_clarify/Solutions/solution_path_to_sdl.lua]: @istoimenova use colors helpers functions

File: to_clarify/Solutions/solution_path_to_sdl.lua (hunk 1 above)
Comment (istoimenova) [inline to_clarify/Solutions/solution_path_to_sdl.lua]: @LuxoftAKutsan, based on previous comment, only declaration here using this function is:
local storagePath = config.SDLStoragePath..commonFunctions:pathJoin(config.application1.registerAppInterfaceParams.appID.. "_", config.deviceMAC)

File: to_clarify/Solutions/solution_path_to_sdl.lua (hunk 3 above)
Comment (istoimenova) [inline to_clarify/Solutions/solution_path_to_sdl.lua]: @LuxoftAKutsan, based on first comment, I think that we should clarify which function is better to be used in this situation: commonPreconditions:GetPathToSDL() or commonFunctions:pathJoin().
And may be the question is more general, not only for relative path verification.

File: to_clarify/Solutions/solution_path_to_sdl.lua (hunk 4 above)
Comment (istoimenova) [inline to_clarify/Solutions/solution_path_to_sdl.lua]: @LuxoftAKutsan, agree with you and when final decision is made I can create such function if this implementation is the solution.

**Comment by istoimenova** [comment]
@aderiabin, did you manage to create solution as discussed?

**Comment by dboltovskyi** [comment]
@istoimenova, From my point of view usage of fuzzy comparison (string.match) in ValidIf () is not a good idea.
I propose another solution: in short - just convert relative path to absolute
In order to do this we can use _readlink_ command
For ex:
```
local function getAbsolutePath(relativePath)
    local commandToExecute = "readlink -f " .. "'" .. relativePath .. "'"
    local db = assert(io.popen(commandToExecute, 'r'))
    local data = assert(db:read('*a'))
    db:close()
    return string.gsub(data, "\n", "")
end
```
This function converts relative path to absolute one
In this way expected path can be constructed in following way:
```
local pathToSDL = getAbsolutePath(config.pathToSDL) .. "/"
local expectedPath = pathToSDL .. "storage/" .. config.application1.registerAppInterfaceParams.appID .. "_" .. config.deviceMAC.. "/" .. "icon.png"
```

**Comment by aderiabin** [comment]
@istoimenova I totaly agree with @dboltovskyi in his solution with function `getAbsolutePath()`

Also additional I propose rewrite function `GetPathToSDL()` using `getAbsolutePath()`

**Comment by istoimenova** [comment]
PR will be closed. Waiting common function from @dboltovskyi.

**Pull request closed without merging**